### PR TITLE
Support for badges on the NavbarItem

### PR DESCRIPTION
**Pull request by bebaoboy**
This PR implements the **badges and dot feature** and integrate them to the navbar_router, with minimal codebase changes.
Based on [badges packagae](https://pub.dev/packages/badges) of flutter.

## Features:
- Customize badges according to [badges packagae](https://pub.dev/packages/badges), including size, color, position, animation,.. and put them into NavbarItem to set up initial badges. Supported const.
```
NavbarItem(Icons.home_outlined, 'Home',
        backgroundColor: colors[0],
        selectedIcon: const Icon(
          Icons.home,
          size: 26,
        ),
        // --- it's here ----
        badge: const NavbarBadge(
          badgeText: "11",
          showBadge: true,
        ))
```
<br>

- Next, you can update your badges by using NavbarNotifier.
`
NavbarNotifier.updateBadges(int index, NavbarBadge badge) // new badges
`
<br>

- Customize your badge with **copyWith**:
`
NavbarBadge.copyWith(...)
`
<br>

- Flexibly show or hide badges on demand
`
NavbarNotifier.makeBadgeVisible(int index, bool visible)
`
<br>

- or set the default behaviour of **hiding on tap**:

-- Setting the attributes in NavbarRouter when you setup your navbar:

`
final bool hideBadgeOnPageChanged;
`

-- Setting it manually anywhere:

`
NavbarNotifier.hideBadgeOnPageChanged = true;
`

## Demo video 
Showing random badges/dots on button click (taking from main.dart in example folder)

https://github.com/maheshmnj/navbar_router/assets/74720131/0e0f68d6-fe31-4976-81ad-27fea8a46118

- Support all types of navbar available (I think so).

*List which issues are fixed by this PR. You must list at least one issue.*
Reference: #36 and #30

## Pre-launch Checklist

- ✅ I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- ✅ I listed at least one issue that this PR fixes in the description above.
- ✅ I updated/added relevant **documentation** (doc comments with `///`).
- ✅ I added **new tests** to check the change I am making, or this PR is [test-exempt].
- ✅ All existing and new tests are passing using `flutter test`
image

P.s: This is my first-time fork so if anything's gone wrong please tell me! Thank you, have a good day! @maheshmnj 

[Contributor Guide]: https://github.com/maheshmnj/searchfield/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by maheshmnj** [comment]
Awesome! @bebaoboy I am currently traveling will review it soon.

**Comment by maheshmnj** [comment]
Meanwhile, can you please add some tests regarding this change? You can refer to existing tests in test folder for your reference. Let me know if you need any help.

**Comment by bebaoboy** [comment]
I'll make some tests when I'm available soon, as I'm still working on my college projects (using the navbar too)
Pretty awesome work!

**Comment by maheshmnj** [comment]
Sure thing take your time, Good luck with your school project. I won't be able to review it for another two days. I need to prepare for my conference too and be back be on this PR soon.

**Comment by bebaoboy** [comment]
Yeah so i made some enhancement and added some badge tests. They're all passed (together with current tests) for all types of navigation bar. 

image

Also, the badges look a bit weird on the default UI for web (default). And umm, this is because I don't know how to customize the background of the navigation bar on web, as of now. Below is the picture, taken from my app:
image

Again, take a look on this PR when you're free 😊 Thank you.

**Comment by maheshmnj** [comment]
> Another thing I want to ask is whether we want to wrap badges over user's child widget of NavbarItem. 

I think having a badge parameter in NavbarItem class would be the right approach.

**Comment by maheshmnj** [comment]
Perhaps we should also take into consideration that parameters in NavbarItems can be updated in runtime.
Otherwise we wrill run into a this issue #38, This is yet to be addressed.

**Comment by maheshmnj** [comment]
@bebaoboy I reviewed your PR its kind of lagging on the web when switching to profile tab, need to investigate further. If you run the app on the web it freezes on switching tabs and the app is no longer usable.

**Comment by bebaoboy** [comment]
Hello, How do I run the app on the web like you said to test it?

**Comment by maheshmnj** [comment]
Strange, I saw severe lag issue last night and now I can't reproduce it. It seems to work pretty smooth. I think we are good on the performance part as of now. I am reviewing the code will share the feedback soon.

**Comment by maheshmnj** [comment]
@bebaoboy I have made some comments can you please review and incorporate them?

Thanks

**Comment by bebaoboy** [comment]
> @bebaoboy I have made some comments can you please review and incorporate them?
> 
> Thanks

Help I don't see them, where are your comments ... Please tell me, thank you

**Review by maheshmnj — CHANGES_REQUESTED** [review]

**Comment by maheshmnj** [comment]
Oops I did not submit the review, Apologies you should see them now.

File: lib/src/navbar_notifier.dart
Comment (bebaoboy) [reply lib/src/navbar_notifier.dart:46]: Ah honestly I still think the result is more or less the same, as copyWith will return another new NavbarBadge() object with the copied properties of the given object 😄

here I just directly update the badge from the function argument. No new object is created.

File: example/pubspec.yaml
Comment (bebaoboy) [reply example/pubspec.yaml:15]: Because the built-in Material Badge is simple and has no animation 😄 so badges package comes in really handy, they have much more customization, and handle appear/disappear animations of badges implicitly.

Plus, it's been showcased on [Package Of The Week](https://youtu.be/_CIHLJHVoN8) by Flutter themselves, so there's no reason to avoid using it, I think.

Comment (maheshmnj) [reply example/pubspec.yaml:15]: Okay sounds good!

File: lib/src/navbar_notifier.dart
Comment (maheshmnj) [reply lib/src/navbar_notifier.dart:46]: We can keep it as it.

File: lib/src/navbar_notifier.dart
Comment (bebaoboy) [reply lib/src/navbar_notifier.dart]: I'm sorry to tell you that this change will break the **hideBadgeOnPageChanged** feature when user tap on an item. 

Moreover, when we set NavbarNotifier.index, we kind of **navigate** to that index, and that's not the situation here. We just want to update the badge, not set the index to that item.

So, I will only keep the singleton.Notify();
image

File: lib/src/navbar_notifier.dart
Comment (bebaoboy) [reply lib/src/navbar_notifier.dart:68]: I split it to a separate function like you said!

File: lib/src/navbar_router.dart
Comment (maheshmnj) [inline lib/src/navbar_router.dart:372]: I think this invocation should be in the else part of 

```dart
if (NavbarNotifier.currentIndex == x) {
```

The badge should hide when user switches to a tab that contains a badge.

e.g if we have 4 tabs and we are on tab 0 and if Tab 1 contains a badge, switching to tab1 should hide the badge lmkwyt.

**Review by maheshmnj — CHANGES_REQUESTED** [review]

File: lib/src/navbar_router.dart
Comment (bebaoboy) [reply lib/src/navbar_router.dart:372]: Hello, but I have tested it and it worked as intended. Also, they can hide the badge on their current tab by tapping on it again too. I don't think theres any reason to change it 😅

File: test/navbar_router_test.dart
Comment (bebaoboy) [reply test/navbar_router_test.dart:701]: Oh, so the tests are the same for each navbar type? I think they're a bit different for each of them so I just put the badgeTest in for each type.
If they're truly the same tests, I'll loop through the types and refractor my badge test!

File: lib/src/navbar_router.dart
Comment (maheshmnj) [reply lib/src/navbar_router.dart:372]: Okay, I tested the app its working as intended, But may I know how are the badges hiding by itself if its running the else part of the code?

**Comment by maheshmnj** [comment]
Rest looks good to me, Once this is merged I will update the documentation here https://github.com/maheshmnj/docs/tree/main/docs/navbar_router

If you would like you are welcome to contribute to the documentation.

File: test/navbar_router_test.dart
Comment (maheshmnj) [reply test/navbar_router_test.dart:701]: Yes kind of similar for each navbar type but,  it would not make any difference if you run the same tests (badeGroupTests) without passing a different navbartype for each of the test

File: lib/src/navbar_notifier.dart
Comment (maheshmnj) [reply lib/src/navbar_notifier.dart]: Sounds good!

File: lib/src/navbar_notifier.dart
Comment (maheshmnj) [inline lib/src/navbar_notifier.dart:48]: You can remove this comment

File: test/navbar_router_test.dart
Comment (bebaoboy) [reply test/navbar_router_test.dart]: Hey, sorry for the delay I was dealing with schoolwork. I've added the code to include more navbar types for badge test and it's still working well! Consider this as resolve 😄

Comment (maheshmnj) [reply test/navbar_router_test.dart]: This looks great!

**Comment by maheshmnj** [comment]
Great work @bebaoboy. Thank you for contribution!
 This feature will be available in the next release.

**Pull request merged**
